File: student_boilerplate/character.js
// character.js  —  YOUR character.
//
// You write JavaScript that returns a character "manifest" object.
// The browser runs this; the server NEVER runs your JS.
//
// Read the Power Manual at /manual for every cast kind and effect.

// ---------- Helper builders --------------------------------------------------
//
// Notice how each helper is just a function that returns an object. Students
// can call these from buildCharacter() with different arguments.

function projectile({ name, key, color, damage, cooldownMs = 700,
                       speed = 480, radius = 6, count = 1, spreadDeg = 0,
                       pierce = false, extraEffects = [] }) {
  return {
    name, key, cooldownMs,
    cast: {
      kind: "projectile", color,
      speed, radius, lifetimeMs: 1800,
      count, spreadDeg, pierce,
      onHit: [{ effect: "damage", amount: damage }, ...extraEffects],
    },
  };
}

function shield({ name = "Shield", key = "e", color = "cyan",
                  amount = 50, durationMs = 3000, cooldownMs = 6000 } = {}) {
  return {
    name, key, cooldownMs,
    cast: { kind: "shield", color, amount, durationMs },
  };
}

// ---------- buildCharacter ---------------------------------------------------

function buildCharacter() {
  // Variables and arrays
  const flameColors = ["yellow", "orange", "red"];

  // For loop + if statement: pick the "hottest" colour.
  let hottest = flameColors[0];
  for (let i = 0; i < flameColors.length; i++) {
    if (flameColors[i] === "orange") {
      hottest = flameColors[i];
    }
  }

  // Build powers programmatically using the helpers above.
  const powers = [];
  powers.push(projectile({
    name: "Fireball", key: "space", color: hottest,
    damage: 18, cooldownMs: 700,
    extraEffects: [{ effect: "dot", dps: 6, durationMs: 1500 }], // burns!
  }));
  powers.push(shield({ name: "Mage Shield", key: "e", amount: 40, durationMs: 2500 }));

  return {
    characterName: "Fire Wizard",
    color: hottest,
    size: 28,
    speed: 220,
    maxHealth: 100,
    powers,
    // Optional: pixel-art sprites (PNG / GIF data: URIs).
    // Use the "Add custom pixel art" widget below the editor to import frames
    // from https://www.piskelapp.com/. Up to 4 frames per slot, 16 KB each.
    // sprites: {
    //   idle:   ["data:image/png;base64,..."],
    //   walk:   ["data:image/png;base64,...", "data:image/png;base64,..."],
    //   attack: ["data:image/png;base64,..."],
    //   hurt:   ["data:image/png;base64,..."],
    // },
  };
}

File: student_boilerplate/character.py
def build_character():
    # Medic — light damage, strong self-sustain.
    # Demonstrates: heal cast, healing area, and a small projectile.

    heal = {
        "name": "Mend",
        "key": "e",
        "cooldownMs": 5000,
        "cast": {
            "kind": "heal",
            "color": "lime",
            "amount": 60,
        },
    }

    # A regen pool: stand in it to heal yourself OR teammates.
    regen_pool = {
        "name": "Regen Pool",
        "key": "q",
        "cooldownMs": 7000,
        "cast": {
            "kind": "area",
            "color": "green",
            "radius": 80,
            "durationMs": 4000,
            "tickIntervalMs": 400,
            # Set followOwner: true if you want the pool to move with you.
            # Default (false) drops it where you cast it.
            "followOwner": True,
            # Note: area effects affect everyone except the caster, so this heals
            # OTHER players in it. Pair with a teammate!
            "onTick": [
                {
                    "effect": "heal",
                    "amount": 6,
                }
            ],
        },
    }

    dart = {
        "name": "Healing Dart",
        "key": "space",
        "cooldownMs": 600,
        "cast": {
            "kind": "projectile",
            "color": "white",
            "speed": 520,
            "radius": 4,
            "lifetimeMs": 1200,
            "count": 1,
            "spreadDeg": 0,
            "onHit": [
                {
                    "effect": "damage",
                    "amount": 10,
                }
            ],
        },
    }

    return {
        "characterName": "Medic",
        "color": "lime",
        "size": 24,
        "speed": 250,
        "maxHealth": 120,
        "powers": [dart, heal, regen_pool],
    }
